### PR TITLE
Added an option to fetch one row at a time.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
* this is a Proof of Concept PR, feedback is welcome
* added a fetch method option to `Query` to support `fetch one row at a time` feature
* this is great for maintenance tasks that are expected to process a lot of data and do not need to more than one record at a time

### CODE REVIEW

File: src/ORM/Connect/PDOQuery.php [hunk 1]
```diff
@@ -18,13 +47,70 @@ class PDOQuery extends Query
 
     protected $results = null;
 
+    /**
+     * @var string
+     */
+    protected $fetchMedthod = self::FETCH_METHOD_ALL;
```
Comment (ScopeyNZ) [inline src/ORM/Connect/PDOQuery.php]: Typo 😉 

```suggestion
    protected $fetchMethod = self::FETCH_METHOD_ALL;
```

Obviously don't accept the suggestion without updating all the other bits of code.

File: src/ORM/Connect/PDOQuery.php [hunk 2]
```diff
@@ -18,13 +47,70 @@ class PDOQuery extends Query
...
+    /**
+     * Invoke a callback which may modify fetch method flag, but ensures this flag is restored
+     * after completion, without modifying global state.
+     *
+     * The desired fetch method flag should be set by the callback directly
+     *
+     * @param callable $callback
+     * @return mixed Result of $callback
+     */
+    public static function withFetchMethodFlag($callback)
```
Comment (ScopeyNZ) [inline src/ORM/Connect/PDOQuery.php]: Wouldn't it make sense to also ask for the intended fetch method as a parameter so this method can handle both changing to a new method and back to the existing method?

File: src/ORM/Connect/PDOQuery.php [hunk 3]
```diff
@@ -34,17 +120,39 @@ public function __construct(PDOStatement $statement)
 
     public function seek($row)
     {
+        // seek is not available for fetch row
+        if ($this->fetchMedthod === static::FETCH_METHOD_ROW) {
+            return [];
```
Comment (ScopeyNZ) [inline src/ORM/Connect/PDOQuery.php]: I believe that throwing an Exception here would actually be more convenient for a developer. Debugging why your row contains no values when you don't realise that seek shouldn't be used with this fetch method would be a footgun. This goes for the other "unsupported" API.

**Comment by mfendeksilverstripe** [comment]
@ScopeyNZ Thanks for feedback. Looking at this code now, I'm thinking that a new class would be a better approach. Will try to revisit this during my PDP time or the next hackday.

**Comment by kinglozzer** [comment]
@mfendeksilverstripe I’m not sure if you’re looking to add this to other drivers too, but just a friendly heads up that we have an RFC open about dropping PDO support entirely https://github.com/silverstripe/silverstripe-framework/issues/8598

**Comment by kinglozzer** [comment]
Actually, it looks like @sminnee already switched non-PDO drivers to use generators for SS5, which should achieve the same goal as this feature: https://github.com/silverstripe/silverstripe-framework/pull/6518. That doesn’t exclude this as a possible SS4 feature, just noting it should be possible in SS5 without any extra flags :D

**Comment by mfendeksilverstripe** [comment]
@kinglozzer Thanks for the update, these are great news. I guess having this option just for SS 4 would still be nice.

**Comment by sminnee** [comment]
Do you know if the problems that this PR fixes occur with the non-PDO driver, @mfendeksilverstripe? Doesn’t block merge, just curious. It would be another validation point for my proposed depreciation of PDO in 4.5

**Comment by mfendeksilverstripe** [comment]
@sminnee Not sure will need to investigate further.

**Comment by mfendeksilverstripe** [comment]
@sminnee @ScopeyNZ I had a look at the https://github.com/silverstripe/silverstripe-framework/pull/6518

I don't that PR fixes the issue with the `PDO` as this driver will still fetch all data at once via this call `$this->statement->fetchAll(PDO::FETCH_NUM)`.

I had a look at other driver - `MySQLi` and that one looks fine. It seems to fetch one row at a time.

I think long term solution is to switch to `MySQLi` so I'm going to close this PR as this was just a `PoC` anyway.

**Pull request closed without merging**